python/ray/spark/utils.py

```diff
@@ -0,0 +1,245 @@
+        returncode=returncode,
+        stdout=stdout,
+        stderr=stderr,
+    )
+    if throw_on_error and returncode != 0:
+        raise ShellCommandException.from_completed_process(comp_process)
+    return comp_process
+
+
+def get_safe_port(ip):
+    import socket
+    """Returns an ephemeral port that is very likely to be free to bind to."""
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    sock.bind((ip, 0))
+    port = sock.getsockname()[1]
+    sock.close()
+    return port
+
+
+def check_port_open(host, port):
+    import socket
+    from contextlib import closing
+    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
+        return sock.connect_ex((host, port)) == 0
+
+
+def get_spark_session():
+    from pyspark.sql import SparkSession
+    return SparkSession.builder.getOrCreate()
+
+
+def get_spark_driver_hostname(spark):
+    spark_master_url = spark.conf.get("spark.master")
+    if spark_master_url.lower().startswith("local"):
+        return "127.0.0.1"
+    else:
+        parsed_spark_master_url = urlparse(spark_master_url)
+        if parsed_spark_master_url.scheme.lower() != "spark" or \
+                not parsed_spark_master_url.hostname:
+            raise ValueError(f"Unsupported spark.master URL: {spark_master_url}")
+        return parsed_spark_master_url.hostname
+
+
+def get_max_num_concurrent_tasks(spark_context):
+    """Gets the current max number of concurrent tasks."""
+    # pylint: disable=protected-access
+    # spark 3.1 and above has a different API for fetching max concurrent tasks
+    if spark_context._jsc.sc().version() >= "3.1":
+        return spark_context._jsc.sc().maxNumConcurrentTasks(
+            spark_context._jsc.sc().resourceProfileManager().resourceProfileFromId(0)
+        )
+    return spark_context._jsc.sc().maxNumConcurrentTasks()
+
+
+def _get_total_phyisical_memory():
+    import psutil
+    return psutil.virtual_memory().total
+
+
+def _get_total_shared_memory():
+    import shutil
+    return shutil.disk_usage('/dev/shm').total
+
+
+def _get_cpu_cores():
+    import multiprocessing
+    return multiprocessing.cpu_count()
+
+
+def get_avail_mem_per_ray_worker(spark):
+    """
+    Return the available heap memory and object store memory for each ray worker.
+    NB: We have one ray node per spark task.
+    """
+    # TODO: add a option of heap memory / object store memory ratio.
+    num_cpus_per_spark_task = int(spark.sparkContext.getConf().get("spark.task.cpus", "1"))
+
+    def mapper(_):
+        try:
+            num_cpus = _get_cpu_cores()
+            num_task_slots = num_cpus // num_cpus_per_spark_task
+
+            physical_mem_bytes = _get_total_phyisical_memory()
+            shared_mem_bytes = _get_total_shared_memory()
+
+            ray_worker_object_store_bytes = int(shared_mem_bytes / num_task_slots * 0.8)
+            ray_worker_heap_mem_bytes = int((physical_mem_bytes - shared_mem_bytes) / num_task_slots * 0.8)
```

Comment:
@ericl 
I propose to add a option `object_store_memory_ratio` for ray on spark , and the rule becomes:

```
available_ray_node_memory = (total_physical_memory / num_spark_task_slots) * 0.8

dev_shm_limitation_per_ray_node = (dev_shm_capacity / num_spark_task_slots) * 0.8

ray_object_store_memory_per_node = min(available_ray_node_memory * object_store_memory_ratio, dev_shm_limitation_per_ray_node)

ray_heap_memory_per_node = available_ray_node_memory - ray_object_store_memory_per_node
```

WDYT ?